test/e2e/suite/nodeadm_test.go

```diff
@@ -348,12 +283,182 @@ var _ = Describe("Hybrid Nodes", func() {
+							)
+
+							nodeKubernetesVersion, err := kubernetes.GetPreviousK8sVersion(test.cluster.KubernetesVersion)
+							Expect(err).NotTo(HaveOccurred(), "expected to get previous k8s version")
+
+							createNodeTest := createNodeTest{
+								awsSession:         test.awsSession,
+								cluster:            test.cluster,
+								ec2ClientV2:        test.ec2ClientV2,
+								instanceName:       instanceName,
+								instanceProfileARN: test.stackOut.InstanceProfileARN,
+								k8sClient:          test.k8sClient,
+								k8sVersion:         nodeKubernetesVersion,
+								logger:             test.logger,
+								nodeadmURLs:        test.nodeadmURLs,
+								os:                 os,
+								provider:           provider,
+								s3Client:           test.s3Client,
+								ssmClient:          test.ssmClient,
+							}
+							instance, err := createNodeTest.Run(ctx)
+							Expect(err).NotTo(HaveOccurred(), "EC2 Instance should have been created successfully")
+
+							joinNodeTest := joinNodeTest{
+								k8s:           test.k8sClient,
+								nodeIPAddress: instance.IP,
+								logger:        test.logger,
+							}
+							Expect(joinNodeTest.Run(ctx)).To(Succeed(), "node should have joined the cluster sucessfully")
+
+							upgradeNodeTest := upgradeNodeTest{
+								k8s:      test.k8sClient,
+								ssm:      test.ssmClient,
+								cluster:  test.cluster,
+								ec2:      instance,
+								logger:   test.logger,
+								provider: provider,
+							}
+							Expect(upgradeNodeTest.Run(ctx)).To(Succeed(), "node should have upgraded successfully")
+							Expect(joinNodeTest.Run(ctx)).To(Succeed(), "node should have joined the cluster sucessfully after nodeadm upgrade")
+
+							test.logger.Info("Resetting hybrid node...")
+
+							uninstallNodeTest := uninstallNodeTest{
+								k8s:      test.k8sClient,
+								ssm:      test.ssmClient,
+								ec2:      instance,
+								provider: provider,
+								logger:   test.logger,
+							}
+							Expect(uninstallNodeTest.Run(ctx)).To(Succeed(), "node should have been reset sucessfully")
+						},
+						Entry(fmt.Sprintf("With OS %s and with Credential Provider %s", os.Name(), string(provider.Name())), context.Background(), os, provider, Label(os.Name(), string(provider.Name()), "upgradeflow")),
+					)
 				}
 			}
 		})
 	})
 })
 
+type createNodeTest struct {
+	awsSession         *session.Session
+	cluster            *peered.HybridCluster
+	ec2ClientV2        *ec2v2.Client
+	instanceName       string
+	instanceProfileARN string
+	k8sClient          *clientgo.Clientset
+	k8sVersion         string
+	logger             logr.Logger
+	nodeadmURLs        e2e.NodeadmURLs
+	os                 e2e.NodeadmOS
+	provider           e2e.NodeadmCredentialsProvider
+	s3Client           *s3v1.S3
+	skipCleanup        bool
+	ssmClient          *ssmv1.SSM
+}
+
+func (c createNodeTest) Run(ctx context.Context) (ec2.Instance, error) {
+	Expect(c.os).NotTo(BeNil())
```

Comment:
I can remove all these expects and just return the err if thats what we want?

Reply:
yeah if the method already returns an error, we should do that.